### ISSUE TITLE
Build and distribute Snappy packages

### ISSUE THREAD

**Issue by Tyriar** [Microsoft/vscode#5458]
Ubuntu 16.04 is going to include a new type of package that installs the application in a more isolated manner.
- https://developer.ubuntu.com/en/desktop/
- https://developer.ubuntu.com/en/snappy/#tour
- http://www.pcworld.com/article/3056668/linux/ubuntu-1604-will-support-snaps-alongside-deb-packages-for-improved-software-installation.html

**Comment by sergiusens** [comment]
Hello there, I have something sort of working.

It is not on the store yet, but you can download it from http://people.canonical.com/~sergiusens/snappy/vscode_1.0_amd64.snap

There are two snappy related bugs still open for this (which are being worked on):

- [LP: #1577514](https://bugs.launchpad.net/snappy/+bug/1577514)
- [LP: #1577520](https://bugs.launchpad.net/snappy/+bug/1577520)

So it would need to be installed by running

    sudo snap install --devmode vscode*.snap

The most important thing that I still need to resolve it plugin installation. That is an area I could use pointers for btw.

I have 4 commits here https://github.com/sergiusens/vscode/commits/release/1.0.0

This branch has:

- a gulp task to create the snap layout.
- snapcraft specific assets.
- a launcher (shell) specific to snapcraft (this won't be necessary for future builds though).
- a local plugin (`parts/plugins/x-nodejs`) until I (or someone else) adds support for either `gulp` as a top level plugin or clean up what I did here to also support gulp.

(a840a34)[https://github.com/sergiusens/vscode/commit/a840a3451061efd0a988799e5383e9382404bd59] you might want to cherry pick regarless.

**Comment by Tyriar** [comment]
Hey @sergiusens, thanks a lot for your work on this. The snap file size is 92mb compared to 30-45 for the other package types, are snaps not compressed by default or was this just left out?

We probably won't delve into distributing a snap until at least [#1577514](https://bugs.launchpad.net/snapcraft/+bug/1577514) is fixed. It sounds like it will be an Electron fix that will then take time for us to uplevel to.

I only thing I can see that's preventing your commits landing in master is moving the build resources (parts, snapcraft.yaml) to `resources/linux/snap` and the build artifacts to `.build/linux/snap`. If you get around to upleveling your branch and making those changes we'd be glad to merge a PR and add it to the CI build.

**Comment by sergiusens** [comment]
@Tyriar ok, sounds like a plan, I'll work on that

**Comment by sergiusens** [comment]
@Tyriar about size, we need to include things in which are not part of the core os which is the common dependency for everything. This could be much smaller but isn't yet as I haven't trimmed down all of it yet.

I'll get to the other items as soon as I look at it. I also based out of `release/1.0.0` which I guess the intenton is to get this into master/devel instead?

**Comment by Tyriar** [comment]
@sergiusens yeah if you update your work to work on master with the tweaks I mentioned then we can push this and I'll get notified if it breaks. It will also make it convenient for anyone who wants to use snappy packages from an OSS build.

**Comment by flexiondotorg** [comment]
@Tyriar Microsoft Visual Studio Code is published as a snap in the store as `vscode` using classic confinement in the beta channel. We'd love for your team to take ownership of this snap and publish new versions directly to the store :-)

The rationale for using classic confinement is that vscode is a code editor and users may reasonably expect to be able to edit files anywhere on the system. Use classic confinement means that [LP: #1577514](http://pad.lv/1577514) is no longer relevant and neither is [LP: #1577520](http://pad.lv/1577520), although the later issue is also now resolved in `snapd` when applications are strictly confined.

Snaps have a number of useful characteristics such as being simple to create, [support multiple Linux distributions](https://snapcraft.io/docs/core/install), enable very simple installation instructions and provide access to install metrics via the [Snap store](https://dashboard.snapcraft.io). There are also various hooks and services [3] available to automate the build and publish pipeline.

**Comment by Tyriar** [comment]
@flexiondotorg thanks for reaching out again, could you email me the process for taking ownership and publishing?

**Comment by Ads20000** [comment]
Update for those interested:
[VS Code now in the snappy store in the stable channel.](https://insights.ubuntu.com/2017/05/19/visual-studio-code-is-now-available-as-a-snap-on-ubuntu/)
Can't be installed from Ubuntu Software...[yet](https://forum.snapcraft.io/t/behaviour-for-installing-classic-devmode-and-unsigned-snaps-in-a-gui/688) (use `sudo snap install vscode --classic` for now), ownership needs to be transferred and it potentially needs to be renamed `code`.

**Comment by vannitotaro** [comment]
Moreover there is no reference to _Microsoft_ in snap detailed information. I think it should be there, to give a hint to the user that it is the right package to install.

```
me@pc:~$ snap info vscode
name:      vscode
summary:   "Code editing. Redefined."
publisher: flexiondotorg
contact:   snappy-canonical-storeaccount@canonical.com
description: |
  Visual Studio Code is a new choice of tool that combines the
  simplicity of a code editor with what developers need for the core
  edit-build-debug cycle.
  
channels:                                 
  latest/stable:    1.12.2-1494422229 (8) 122MB classic
  latest/candidate: 1.12.2-1494422229 (8) 122MB classic
  latest/beta:      1.12.2-1494422229 (8) 122MB classic
  latest/edge:      1.12.2-1494422229 (8) 122MB classic
```

**Comment by Ads20000** [comment]
@vannitotaro once Martin has transferred ownership to Microsoft, it will hopefully read
`publisher: microsoft`

**Comment by cossio** [comment]
Any news on this? Do we still need the `--classic`?

**Comment by Ads20000** [comment]
Classic is still needed and Martin has yet to transfer owner to Microsoft

**Comment by Tyriar** [comment]
Update from my side is that I've been a little stuck on the ownership transfer as I haven't had enough time to get around to setting this up.

**Comment by phillipsj** [comment]
Any updates?  This would be awesome to have.

**Comment by Tyriar** [comment]
I'll be working on this in a couple of weeks, expect another update soon 😃

**Comment by ghisvail** [comment]
The current snap for `vscode` still lists Martin as publisher and is lagging behind upstream (Snap version is `1.15.1`, current is `1.16.1`). Is there anything else preventing the snap package to be updated to the latest version besides the transfer of ownership to Microsoft?

**Comment by Tyriar** [comment]
@ghisvail the new snap name is `code` to be consistent with the CLI. I'm currently at Ubuntu Rally with @flexiondotorg and the rest of the team working on the snap package build and integrating it into our CI.

**Comment by ghisvail** [comment]
@Tyriar Oh, I heard about the rally. So great your team is collaborating with Canonical directly.

**Comment by Tyriar** [comment]
PR to build snap packages: https://github.com/Microsoft/vscode/pull/35317

---

To build it in our CI we would need to run:

```
apt install -y snapcraft
export SNAPCRAFT_SETUP_CORE=1
export LC_ALL=C.UTF-8
export LANG=C.UTF-8
```

And then

```
npm run gulp -- "vscode-linux-$ARCH-build-snap"
```

**Comment by ghisvail** [comment]
I saw the PR was merged. Does it mean we'll be getting an official snap package from the next release onwards then?

**Comment by Tyriar** [comment]
We need to verify some legal stuff and get it integrated into our CI, expect movement on this in the next 1-2 months 😃

**Comment by Nikki1993** [comment]
@Tyriar any chance to see it in November release? Solus is about to drop the vscode from main repos since VSCode still requires npm4 to be built. Been using tar.gz for a while now, a bit of an inconvenience with lack of automatic PATH and updates.

**Comment by karenachiketc** [comment]
@Tyriar any updates on the snap package. I'm a bit sceptical to install this
```
$ snap info vscode
name:      vscode
summary:   Code editing. Redefined.
publisher: Martin Wimpress
contact:   snappy-canonical-storeaccount@canonical.com
description: |
  Visual Studio Code is a new choice of tool that combines the
  simplicity of a code editor with what developers need for the core
  edit-build-debug cycle.
snap-id:     XPQdduIwHiDCZvPHRrmsqV7Nr6nQRuqM
channels:                           
  stable:    1.19.1-1513676564 (23) 134MB classic
  candidate: 1.19.1-1513676564 (23) 134MB classic
  beta:      1.19.1-1513676564 (23) 134MB classic
  edge:      1.19.1-1513676564 (23) 134MB classic
```

**Comment by ghisvail** [comment]
@karenachiketc Martin is a canonical employee and a skilled snapcrafter. I would not doubt of the quality of this (I used it before testing the flatpak version). You may request him to update it to the latest version though.

**Comment by karenachiketc** [comment]
@ghisvail No not all. I'm not doubting Martin's quality of work here. 
I apologize for not making this clear in my first comment: I was sceptical because VS code is a product by Microsoft and that's what the snap should reflect. I would be much more comfortable to install it then :)

**Comment by Ads20000** [comment]
@karenachiketc I'm assuming this is temporary, the snappy team want upstreams to hook snappy builds into their CI and distribute their snaps themselves. @flexiondotorg may have more on this...

**Comment by Tyriar** [comment]
I was just at the snapcraft summit @flexiondotorg and co and made some good progress, expect news soon. In the meantime I would recommend @flexiondotorg publish 1.19.3 based on the deb package to his snap as it includes a security fix.

**Comment by karenachiketc** [comment]
Thank you for the update :)
I thought the snap was not based on the deb packages(available on the VS code website) and was developed independently by @flexiondotorg
I guess I'll go ahead and install the snap after it has been updated the mentioned version.

**Comment by Tyriar** [comment]
There was a report that the system IME doesn't work in the snap package https://github.com/Microsoft/vscode/issues/43881

**Comment by DarkLite1** [comment]
Using the current table snap package, but it's not up to date:

```
[me@machine~] sudo snap info vscode 
name:      vscode
summary:   Code editing. Redefined.
publisher: flexiondotorg
contact:   snappy-canonical-storeaccount@canonical.com
license:   unknown
description: |
  Visual Studio Code is a new choice of tool that combines the
  simplicity of a code editor with what developers need for the core
  edit-build-debug cycle.
snap-id: XPQdduIwHiDCZvPHRrmsqV7Nr6nQRuqM
commands:
  - vscode
tracking:    stable
installed:   1.20.1-1518535978 (27) 80MB classic
refreshed:   2018-03-06 11:06:14 +0100 CET
channels:                           
  stable:    1.20.1-1518535978 (27) 80MB  classic
  candidate: 1.20.1-1518535978 (27) 80MB  classic
  beta:      1.20.1-1518535978 (27) 80MB  classic
  edge:      1.21.1-1521038896 (29) 127MB classic
```

When checking within VSCode, Help > Check for updates, it's cleat there's a new build available. However, when trying to update the snap it's not there:

```
[me@machine~] sudo snap refresh vscode
snap "vscode" has no updates available
```
Can someone check why the snap package is not updated? Thank you.

**Comment by Ads20000** [comment]
Looks like they haven't released the 1.21.1 snap to `stable` yet, try `snap refresh vscode --edge --classic`, be warned that it may not work and you'll want to use `snap refresh vscode --stable classic` when it's in stable, otherwise you may be released to a development version of Code without you realizing!

**Comment by DarkLite1** [comment]
@Tyriar how is going with the integration in CI regarding the legal matters? The fix for the IME bug seems to be merged.

**Comment by Tyriar** [comment]
@DarkLite1 unfortunately we got blocked by a snap issue (which is now resolved) and other priorities took over. It's on hold currently as we enter endgame for the March release.

**Comment by chyzwar** [comment]
can we somehow help with snap package?

**Comment by Tyriar** [comment]
Update: I've had too much other work to focus on to take this to the finish line. One of the things that was really bugging me is that every time I went to get some work done on it I would run into another major issue that needed help from the snapcraft team like random crashes or things not working.

> can we somehow help with snap package?

Potentially, here's the branch: https://github.com/Microsoft/vscode/tree/tyriar/snap

I think this is a complete list of current work left to do:

- [ ] Make sure update flow works; the update service needs to scan the relevant `/snap` folder to check for an update. Most of this is in place on the branch but I don't think it's working just yet https://github.com/Microsoft/vscode/compare/tyriar/snap
- [ ] Figure out the crash/instant logout that occurs when launching on some 17.10 machines
- [ ] Integrate into our CI

**Comment by DarkLite1** [comment]
Thank you for working on the snap @Tyriar , I've just received the following update:
```
sudo snap list
vscode                  1.23.0-1525361119  34    stable    flexiondotorg  classic
```
So it seems like the integration into the CI is working now, unless someone else packaged a new snap manually.

**Comment by ghisvail** [comment]
flexiondotorg is Martin Wimpress from Canonical.

**Comment by Tyriar** [comment]
Yes that's not the one published by us, @flexiondotorg derives that one from the deb

**Comment by tomByrer** [comment]
At this time, [the snap package](https://snapcraft.io/vscode) is stuck at v1.24 while the main distos have had v1.25 for four days now.  I know this is a US holiday weekend, but unfortunately my ]usual distro](https://aur.archlinux.org/packages/code-git/) crashes since yarn won't build there.

**Comment by Ads20000** [comment]
@flexiondotorg

**Comment by Tyriar** [comment]
I reopened issues related to snap package and labelled them here: https://github.com/Microsoft/vscode/labels/install-update-snap

**Comment by joaomoreno** [comment]
This is now in `master`, planning to release for March.

**Comment by mightyiam** [comment]
Do you mean April, @joaomoreno?

**Comment by joaomoreno** [comment]
Actually, March! 🎆
